### PR TITLE
Filter exact list prefix matches for azure gen2 accounts

### PR DESCRIPTION
**Pull request by roeap**
# Which issue does this PR close?

Part of https://github.com/apache/arrow-rs/issues/3712

again, no straight forward way to test this in CI without a gen2 account, but tested it locally.

# Rationale for this change
 

# What changes are included in this PR?

# Are there any user-facing changes?

### CODE REVIEW

**Comment by roeap** [comment]
I know the latest object store release is just a week old, but there are some fixes and features on master I am keen to integrate. Would it be possible to cut another object store release soon?

File: object_store/src/azure/client.rs [hunk 1]
```diff
@@ -419,33 +419,37 @@ struct ListResultInternal {
...
+        .blobs
+        .blobs
+        .into_iter()
+        .map(ObjectMeta::try_from)
+        // Note: workaround for gen2 accounts with hierarchical namespaces. These accounts also
+        // return path segments as "directories" and include blobs in list requests with prefix,
+        // if the prefix mateches the blob. When we want directories, its always via
+        // the BlobPrefix mechanics, and during lists we state that prefixes are evaluated on path segement basis.
+        .filter_map_ok(|obj| {
+            if obj.size > 0 && obj.location != prefix {
```
Comment (tustvold) [inline object_store/src/azure/client.rs:441]:
```suggestion
            if obj.size > 0 && obj.location.as_ref().len() > prefix.as_ref().len() {
```
Should be cheaper and equivalent

**Review by tustvold — APPROVED** [review]

**Comment by tustvold** [comment]
> Would it be possible to cut another object store release soon?

I can look to cut a release once this is in

**Comment by roeap** [comment]
I seem to not be able to re-start the failing tests, but to me it seems is just a bit of flakiness in connecting to localstack.

**Pull request merged**